gluten-core/src/main/scala/org/apache/spark/util/SparkDirectoryUtil.scala

```diff
@@ -67,16 +68,19 @@ class SparkDirectoryUtil private (val roots: Array[String]) extends Logging {
 }
 
 object SparkDirectoryUtil extends Logging {
+  private val state = new AtomicBoolean(false)
   private var INSTANCE: SparkDirectoryUtil = _
 
-  def init(conf: SparkConf): Unit = synchronized {
+  def init(conf: SparkConf): Unit = {
     val roots = Utils.getConfiguredLocalDirs(conf)
     init(roots)
   }
 
-  private def init(roots: Array[String]): Unit = synchronized {
+  private def init(roots: Array[String]): Unit = {
     if (INSTANCE == null) {
-      INSTANCE = new SparkDirectoryUtil(roots)
+      if (state.compareAndSet(false, true)) {
+        INSTANCE = new SparkDirectoryUtil(roots)
+      }
       return
```

Comment:
It's not the hot code path, let's remove the `if (INSTANCE == null)` condition.

Just directly call  `compareAndSet`.

Reply:
Because remove `if (INSTANCE == null)` will cause NPE. So I removed `if (INSTANCE.roots.toSet != roots.toSet)`.